### DISCUSSION TITLE
embedding dimension mismatch when enable use_cvm=True

### DISCUSSION THREAD

**Discussion by kiminh** [Qihoo360/tensornet#70] [General] [answered]
model structure: wide and deep; deep part: 31 single-value feature slots, dimension:8; linear part: 31 single-value feature slots, dimension:1
```
def create_sub_model(linear_embs, deep_embs, deep_hidden_units):
    linear_emb_input_shapes = [emb.shape for emb in linear_embs]
    deep_emb_input_shapes = [emb.shape for emb in deep_embs]

    linear_inputs = [
        tf.keras.layers.Input(name="linear_emb_{}".format(i), dtype="float32", shape=shape[1:])
        for i, shape in enumerate(linear_emb_input_shapes)
    ]
    deep_inputs = [
        tf.keras.layers.Input(name="deep_emb_{}".format(i), dtype="float32", shape=shape[1:])
        for i, shape in enumerate(deep_emb_input_shapes)
    ]

    linear, deep = None, None

    if linear_inputs:
        linear = tf.keras.layers.Concatenate(name="linear_concact", axis=-1)(linear_inputs)

    if deep_inputs:
        deep = tf.keras.layers.Concatenate(name="deep_concact", axis=-1)(deep_inputs)

        for i, unit in enumerate(C.DEEP_HIDDEN_UNITS):
            deep = tf.keras.layers.Dense(unit, activation="relu", name="dnn_{}".format(i))(deep)

    both = tf.keras.layers.concatenate([deep, linear], name="both")
    both = tn.layers.TNBatchNormalization(synchronized=True, sync_freq=4, max_count=1000000)(both)
    output = tf.keras.layers.Dense(1, activation="sigmoid", name="pred")(both)

    return tn.model.Model(inputs=[linear_inputs, deep_inputs], outputs=output, name="sub_model")
```
the input dimension of dnn is 310, which is 31 * 10 (embedding dimension here is 10, it should be the (show, log(click)-log(show)) when use_cvm_=True), while in the sparse table output the dimension is 8. So should I manually append the two dimension with embedding output in sparse table when serving, or is that a bug should be fixed?  @thundercatli  @gzm55 

https://github.com/Qihoo360/tensornet/blob/8950fe456b230a7b061e8a8a784dfd3f8c2c9d4b/core/ps/optimizer/ada_grad_kernel.cc#L123
```
    if (opt->ShouldUseCvm()) {
        float log_show = log(show_ + 1);
        float log_click = log(click_ + 1);
        w[dim] = show_;
        w[dim + 1] = (log_click - log_show);
    }
```

![image](https://github.com/user-attachments/assets/0e26e1f2-6e25-4bd5-9e05-075e49a3e097)

sparse_ada_grad_schema = ["sign", "dim", "weights", "g2sum", "show", "no_show_days", "click"]

https://github.com/Qihoo360/tensornet/blob/master/core/ps/optimizer/ada_grad_kernel.cc#L131C1-L143C2
```
void SparseAdaGradValue::SerializeTxt_(std::ostream& os, int dim) {
    os << dim << "\t";
    for (int i = 0; i < dim; i++) {
        os << Weight()[i] << "\t";
    }

    os << g2sum_ << "\t";
    os << show_ << "\t";
    os << no_show_days_ << "\t";
    if (use_cvm_) {
        os << click_;
    }
}
```

wide_input
```
5019009485037015500     8       -0.0251741      0.0863645       -0.0471388      -0.0125073      0.010965        0.00853542      0.0548797  0.0100905        0.1     4.9     0       0
1729890886537794504     8       -0.0315955      -0.00701615     -0.0104996      -0.0018675      -0.0897293      -0.00370546     0.0172712  -0.0493035       0.1     1.96    0       0
9101897099276267728     8       0.0670728       -0.0234302      -0.0319534      -0.0207474      0.0126464       -0.00947607     0.0121802  -0.0833604       0.1     3.92    0       0
4084354851502254628     8       0.0178514       -0.0137977      0.0187439       0.0255161       0.0217853       -0.0553451      -0.0185358 6.97765e-05      0.1     3.92    0       0
5067706786636464832     8       0.0328278       -0.0386421      0.00360148      -0.0230908      -0.0595129      0.007578        0.0191766  -0.0409198       0.1     0.98    0       0
8356821084489907720     8       -0.021308       0.0406726       -0.0401112      -0.085917       0.0259272       0.0276935       0.0282589  -0.0109492       0.1     1.96    0       0
8658039581685739696     8       0.0112502       0.0116276       0.0215849       0.037315        0.00201444      -0.0342867      0.0234399  0.0312791        0.1     0.98    0       0
366730011146381060      8       -0.0367857      0.0671851       -0.00364767     -0.0463901      -0.0181148      -0.063604       0.0379545  -0.059832        0.1     1.96    0       0
```

linear_input
```
5019009485037015500     1       0.029987        0.1     4.9     0       0
1729890886537794504     1       0.0971708       0.1     1.96    0       0
9101897099276267728     1       -0.0887944      0.1     3.92    0       0
4084354851502254628     1       0.00561151      0.1     3.92    0       0
5067706786636464832     1       -0.243389       0.1     0.98    0       0
8356821084489907720     1       -0.0871344      0.1     1.96    0       0
8658039581685739696     1       -0.0236446      0.1     0.98    0       0
366730011146381060      1       -0.0129427      0.1     1.96    0       0
```

**Comment by thundercatli** [comment, marked as answer]
You should append show, click from sparse output for serving. And cvm is a optional plugin, it's not indispensable. It's not 100% positive for model training.

**Reply by kiminh** [reply]
> You should append show, click from sparse output for serving. And cvm is a optional plugin, it's not indispensable. It's not 100% positive for model training.

thanks for your reply. I will do more experiments on whether using cvm plugin.
